src/main/kotlin/io/github/facilityapi/intellij/fsd.bnf

```diff
@@ -42,7 +42,7 @@
 
 definition ::= (attribute_list* service_spec) markdown_remarks*
 
-service_spec ::= service identifier service_items { pin=1 }
+service_spec ::= (service identifier service_items[';']) { pin=1 }
```

Comment:
```diff
--- a/src/main/kotlin/io/github/facilityapi/intellij/fsd.bnf
+++ b/src/main/kotlin/io/github/facilityapi/intellij/fsd.bnf
@@ -42,8 +42,7 @@
 
 definition ::= (attribute_list* service_spec) markdown_remarks*
 
-service_spec ::= service identifier service_items { pin=1 }
-service_items ::= ('{' decorated_service_item* '}') | decorated_service_item
+service_spec ::= (service identifier ';' decorated_service_item*) | (service identifier '{' decorated_service_item* '}') { pin=1 }
```
Will probably need something like this.

You get a good deal of control around how IntelliJ should handle invalid parser states, which is helpful for error messages in the editor. It definitely had some room for improvement before, but it's something else you might consider. That's what the pin and recoverWhile bits are about.